neqo-bin/src/server/mod.rs

```diff
@@ -224,36 +225,57 @@ impl ServerRunner {
         }
     }
 
-    /// Tries to find a socket, but then just falls back to sending from the first.
-    fn find_socket(&mut self, addr: SocketAddr) -> &mut crate::udp::Socket {
-        let ((_host, first_socket), rest) = self.sockets.split_first_mut().unwrap();
-        rest.iter_mut()
-            .map(|(_host, socket)| socket)
-            .find(|socket| {
-                socket
-                    .local_addr()
-                    .ok()
-                    .map_or(false, |socket_addr| socket_addr == addr)
-            })
-            .unwrap_or(first_socket)
-    }
-
-    async fn process(&mut self, mut dgram: Option<Datagram>) -> Result<(), io::Error> {
+    async fn process(&mut self, mut ready_socket_index: Option<usize>) -> Result<(), io::Error> {
+        let mut input_dgrams: Option<DatagramIter> = None;
         loop {
-            match self.server.process(dgram.take(), (self.now)()) {
+            let input_dgram = if let Some(d) = input_dgrams.iter_mut().flatten().next() {
```

Comment:
I found this code to be a little unintuitive.  You are taking a mutable iterator over the option, then flattening it.  It's not clear that you are mutating the underlying iterator as a result of calling `next()`.

Taking a step back, I think that this code is fairly simple:

You read from the indicated socket, process every datagram that it produces, and stop when the socket stops producing datagrams.

Would this structure work?

```rust
fn whatever(ready_socket_index: Option<usize>) -> Res<()> {
    let Some(inx) = ready_socket_index else {
        return Ok(());
    };

    let (host, socket) = self.sockets.get_mut(inx).unwrap();
    while let Some(input_dgrams) = socket.recv(*host, &mut self.recv_buf)? {
        for input in input_dgrams {
            match self.server.process(input, (self.now)()) {
                // see below for a note about sending.
                Output::Datagram(output) => Self::send(&mut self.sockets, &output).await?,
                Output::Callback(t) => {
                    self.timeout = Some(Box::pin(tokio::time::sleep(new_timeout)))
                }
                Output::None => {}
            }
        }
    }
    Ok(())
}
```

I think that's functionally equivalent to what you have, but a lot easier to read, at least for me.

I'm not a borrow checker, so I couldn't say if this works.  There are a lot of references being held here.  I can't see any obvious overlap.

Reply:
> I found this code to be a little unintuitive.

Agreed. The complexity stems from the fact that `neqo_transport::Server` does not have a `process_multiple_input` function, thus having to handle individual output `Datagram`s while still buffering input `Datagram`s.

`neqo_transport::Server` does not have a `process_multiple_input` function, because the set of input `Datagram`s provided might each be for a different `neqo_transport::Connection`, thus each result in a `Output::Datagram` and thus `process_multiple_input` would need to return a set of output `Datagram`s, not just one `Datagram`.

If you think it is worth it, I can explore this pathway, i.e. adding `process_multiple_input` to `neqo_transport::Server`. Preferably in a follow-up pull request.

---

Addressing the concrete suggestion above:

``` rust
    let Some(inx) = ready_socket_index else {
        return Ok(());
    };
```

`process` (or `whatever` above) might be called with `ready_socket_index` `None`, in which case it is expected to drive the output path only, i.e. not just `return` early.

```
    while let Some(input_dgrams) = socket.recv(*host, &mut self.recv_buf)? {
        for input in input_dgrams {
            match self.server.process(input, (self.now)()) {
                // see below for a note about sending.
                Output::Datagram(output) => Self::send(&mut self.sockets, &output).await?,
                Output::Callback(t) => {
                    self.timeout = Some(Box::pin(tokio::time::sleep(new_timeout)))
                }
                Output::None => {}
            }
        }
    }
```

If `self.server.process` returns `Output::Datagram`, one has to call it again, until it returns `Output::Callback` or `Output::None`. In the above suggestion, `self.server.process` is only called, if more input datagrams are available.

Reply:
It seems like there are two functions here then:

1. When `ready_socket_index` is `None`, where you just call `process_output` and send out what comes.
2. When `ready_socket_index` is `Some`, where you do what I outlined.

That leads to:

```rust
    /// Returns `true` if it wants to be called again.
    fn drive_sending(&mut self, mut input: Option<Datagram>) {
        loop {
            match self.server.process(input.take(), (self.now)()) {
                // see below for a note about sending.
                Output::Datagram(output) => Self::send(&mut self.sockets, &output).await?,
                Output::Callback(t) => {
                    self.timeout = Some(Box::pin(tokio::time::sleep(new_timeout)));
                    return;
                }
                Output::None => return,
            }
            true
        }
    }

    let Some(inx) = ready_socket_index else {
        self.drive_sending(None);
        return Ok(());
    };

    while let Some(input_dgrams) = socket.recv(*host, &mut self.recv_buf)? {
        for input in input_dgrams {
            drive_sending(input);
        }
    }
```

Reply:
If I correctly understand the above, https://github.com/mozilla/neqo/pull/2202/commits/91c7920fd282dbcf12839f2d8b2c1051e6c78956 implements your suggestion. Please take a look @martinthomson.

> ``` rust
> fn drive_sending(&mut self, mut input: Option<Datagram>) {
> ```

Note that when calling `drive_sending` with an input datagram, one can not pass `&mut self` to `drive_sending`, as `self.recv_buf` is borrowed already.